<launch>
    <arg name="use_sim_time" default="true"/>
    <arg name="log_level" default="warn"/>

    <arg name="src_image" default="/sensing/camera/traffic_light/image_raw/compressed"/>
    <arg name="src_info" default="/sensing/camera/traffic_light/camera_info"/>
    <arg name="resized_image" default="/sensing/camera/undistorted/image_raw"/>
    <arg name="resized_info" default="/sensing/camera/undistorted/camera_info"/>

    <arg name="min_segment_length" default="-1.0"/>
    <arg name="max_segment_distance" default="-1.0"/>

    <arg name="truncate_pixel_threshold" default="-1" description="For robustness verification, if positive value is given, line segments are truncated."/>

    <!-- ========================================= -->
    <!-- ============ undistort & resize ========= -->
    <!-- ========================================= -->
    <node name="undistort" pkg="vector_map_visual_localizer" exec="undistort_node" output="screen" args="--ros-args --log-level $(var log_level)">
        <param name="use_sim_time" value="$(var use_sim_time)"/>
        <param name="width" value="800"/>
    </node>

    <!-- ========================================= -->
    <!-- ========== line segment detector ======== -->
    <!-- ========================================= -->
    <node name="lsd" pkg="vector_map_visual_localizer" exec="lsd_node" output="screen" args="--ros-args --log-level $(var log_level)">
        <param name="use_sim_time" value="$(var use_sim_time)"/>
        <remap from="$(var src_image)" to="$(var resized_image)"/>
        <remap from="$(var src_info)" to="$(var resized_info)"/>
    </node>

    <!-- ========================================= -->
    <!-- ======== graph based segmentation ======= -->
    <!-- ========================================= -->
    <node if="true" name="graph_segment" pkg="vector_map_visual_localizer" exec="graph_segment_node" output="screen" args="--ros-args --log-level info">
        <remap from="$(var src_image)" to="$(var resized_image)"/>
        <remap from="$(var src_info)" to="$(var resized_info)"/>
    </node>

    <!-- ========================================= -->
    <!-- ============= segment fitler ============ -->
    <!-- ========================================= -->
    <node if="true" name="segment_filter" pkg="vector_map_visual_localizer" exec="segment_filter_node" output="screen" args="--ros-args --log-level info">
        <remap from="$(var src_image)" to="$(var resized_image)"/>
        <remap from="/input/camera_info" to="$(var resized_info)"/>
        <param name="truncate_pixel_threshold" value="$(var truncate_pixel_threshold)" />
        <param name="min_segment_length" value="$(var min_segment_length)"/>
        <param name="max_segment_distance" value="$(var max_segment_distance)"/>
    </node>

</launch>
